### ISSUE TITLE
Get Configuration From Startup Class

### ISSUE THREAD

**Issue by brent-russell** [giggio/FluentMigrator.Runner.DNX#3]
It would be great if the integrated command line runner could grab the configuration from the Startup class like the EF migration tool does (or rather is supposed to - because it's still not fixed as of beta5).

The EF runner commands have options to specify a DbContext class and the startup project name. From there it's supposed to essentially initialize the Startup class to build the configuration and service dependencies, which in turn should configure the runner.

Some way to point the FM runner to a configured connection string would be ideal.

**Comment by giggio** [comment]
So the idea is to get the connection string from the configuration, right? From `Data:DefaultConnection:ConnectionString`?
Also, what if the user wants to use a different connection string?
Maybe we could add a parameter like `--from-config` as it is not always that this configuration is there. I am, for instance, using a migrations project that is separated from my app project, so there is no configuration there.
What do you think?

**Comment by brent-russell** [comment]
Yeah.

The configuration functionality included in the framework is really powerful, and where the configuration comes from is defined in the Startup class. The common practice is to have one or more config.\* files in the project (could be .json, could be .ini), but the values in those files will likely be placeholders, and the values that should be used on any given machine will override those values from either the UserSecrets store or environment variables. The precedence of these and the types of files are defined in the Startup class.

So a typical configuration workflow is:
1. Gather all configuration files and add them to the configuration object
2. Gather all UserSecrets values and add them to the configuration object
3. Gather all environment variables and add them to the configuration object.

So as a developer or in any other environment, I can set a connection string in my user secrets for the project, or an environment variable and that would override the config value from the file.

I think it comes down to pointing the runner to both the migrations and the startup project, so it can leverage the configuration framework of .NET as well as the configuration structure defined by the project.

FM already supports a command line parameter for connection string, and I think that should stay and be the final override. It may need a connection string name or connection string config key parameter in case it's defined in a different config structure.

FM also already has a command line parameter to point to the assembly where migrations are stored. EF has a similar parameter where you can specify a DbContext to use even if it's not in the same project the command is being run from.

EF also has a parameter where you can specify the startup project. That handles cases when the migrations are in a separate class library. There should either be a runnable project or at least a test project in the solution if migrations are in a class library.

**Comment by brent-russell** [comment]
So if you run the FM runner from the startup project, you would have to specify the migration project/assembly location. If you run it from the migrations project, you would have to specify either the startup project or the connection string.

**Comment by giggio** [comment]
It makes sense to me.

**Comment by lenardg** [comment]
We actually used to create a "dummy" startup class into the project with the migration. That way it can be used/deployed without the main program to run migrations. The dummy startup class can then just read user secrets and environment variables, that way it can be used locally or in say Azure web sites, where settings come from environment variables.

Also, by specifying to use the same user secrets for both the main project and the one with the migrations you do not even need separate configurations.

I am just saying this so that it does not need to be restricted that the migration and main project are different, the migration project can be the one to run directly.

**Comment by giggio** [comment]
@lenardg I agree that we should allow for the projects to be separated or not.
The problem of creating a configuration without the main project's config is that we do not know the config source the project is using. We cannot assume it is the default, they might be reading some file we did not predict, for example.
